pkg/reconciler/workload/translate.go

```diff
@@ -0,0 +1,206 @@
+			objs, err := t(ctx, w)
+			if err != nil {
+				return nil, err
+			}
+			for _, wrap := range wp {
+				if objs, err = wrap(ctx, w, objs); err != nil {
+					return nil, err
+				}
+			}
+			return objs, nil
+		},
+	}
+}
+
+// A TranslateFn translates a workload into an object.
+type TranslateFn func(context.Context, Workload) ([]Object, error)
+
+// Translate workload into object or objects with no wrappers.
+func (fn TranslateFn) Translate(ctx context.Context, w Workload) ([]Object, error) {
+	return fn(ctx, w)
+}
+
+var _ Translator = TranslateFn(NoopTranslate)
+
+// NoopTranslate does not translate the workload and does not return error.
+func NoopTranslate(ctx context.Context, w Workload) ([]Object, error) {
+	return nil, nil
+}
+
+// A TranslationWrapper wraps the output of a workload translation in another
+// object or adds addition object.
+type TranslationWrapper func(context.Context, Workload, []Object) ([]Object, error)
+
+var _ TranslationWrapper = NoopWrapper
+
+// NoopWrapper does not wrap the workload translation and does not return error.
+func NoopWrapper(ctx context.Context, w Workload, objs []Object) ([]Object, error) {
+	return objs, nil
+}
+
+var _ TranslationWrapper = KubeAppWrapper
+
+// KubeAppWrapper wraps a set of translated objects in a KubernetesApplication.
+func KubeAppWrapper(ctx context.Context, w Workload, objs []Object) ([]Object, error) {
+	if objs == nil {
+		return nil, nil
+	}
+
+	app := &workloadv1alpha1.KubernetesApplication{}
+
+	for _, o := range objs {
+		u, err := runtime.DefaultUnstructuredConverter.ToUnstructured(o)
+		if err != nil {
+			return nil, errors.Wrap(err, errWrapInKubeApp)
+		}
+
+		kart := workloadv1alpha1.KubernetesApplicationResourceTemplate{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: fmt.Sprintf("%s-%s", o.GetName(), strings.ToLower(o.GetObjectKind().GroupVersionKind().Kind)),
+				Labels: map[string]string{
+					labelKey: string(w.GetUID()),
+				},
+			},
+			Spec: workloadv1alpha1.KubernetesApplicationResourceSpec{
+				Template: &unstructured.Unstructured{Object: u},
+			},
+		}
+
+		app.Spec.ResourceTemplates = append(app.Spec.ResourceTemplates, kart)
+	}
+
+	app.SetName(w.GetName())
+
+	app.Spec.ResourceSelector = &metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			labelKey: string(w.GetUID()),
+		},
+	}
+
+	return []Object{app}, nil
+}
+
+var _ TranslationWrapper = ServiceInjector
+
+// ServiceInjector adds a Service object for every Deployment that has a
+// container with port defined into a workload translation.
+func ServiceInjector(ctx context.Context, w Workload, objs []Object) ([]Object, error) {
+	if objs == nil {
+		return nil, nil
+	}
+
+	for _, o := range objs {
+		if o.GetObjectKind().GroupVersionKind() != deploymentGroupVersionKind {
+			continue
+		}
+
+		d, ok := o.(*appsv1.Deployment)
+		if !ok {
+			return nil, errors.New(errInjectService)
+		}
+
+		for _, c := range d.Spec.Template.Spec.Containers {
```

Comment:
I don't think so, good catch!

Reply:
To be clear, currently, the `ServiceInjector` creates a `Service` for each `Container` defined. I feel like that makes more sense than just picking the first `Container`. This does need to be updated to avoid naming collisions if we continue along this path though.